### PR TITLE
make PipelineModelWrapper trainable

### PR DESCRIPTION
**Pull request by ys-teh**
# ALCHEMI Toolkit Pull Request

## Description

Update PipelineModelWrapper so it can be used in training mode.

## Type of Change

- [x] Bug fix (non-breaking change that fixes an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Performance improvement
- [ ] Documentation update
- [ ] Refactoring (no functional changes)
- [ ] CI/CD or infrastructure change

## Changes Made

- Update PipelineModelWrapper when self.training=True and grad is enabled.
- Add MockTrainableAutogradEnergyModel and test_training_preserves_force_graph_for_backward for testing.

## Testing

- [x] Unit tests pass locally (`make pytest`)
- [x] Linting passes (`make lint`)
- [ ] New tests added for new functionality meets coverage expectations?

## Checklist

- [x] I have read and understand the [Contributing Guidelines](../CONTRIBUTING.md)
- [ ] I have updated the [CHANGELOG.md](../CHANGELOG.md)
- [ ] I have performed a self-review of my code
- [ ] I have added docstrings to new functions/classes
- [ ] I have updated the documentation (if applicable)

## Additional Notes

> [!TIP]
> This repository uses Greptile, an AI code review service, to help conduct
> pull request reviews. We encourage contributors to read and consider suggestions
> made by Greptile, but note that human maintainers will provide the necessary
> reviews for merging: Greptile's comments are **not** a qualitative judgement
> of your code, nor is it an indication that the PR will be accepted/rejected.
> We encourage the use of emoji reactions to Greptile comments, depending on
> their usefulness and accuracy.

### CODE REVIEW

File: test/models/test_pipeline.py
Comment (laserkelvin) [inline test/models/test_pipeline.py:1]: Can you add one or two cases to the unit test cases that will include stress, and/or some scenario where you might need to use different ("non-standard"?) autograd input/outputs to train?

File: nvalchemi/models/pipeline.py
Comment (laserkelvin) [inline nvalchemi/models/pipeline.py:939]: I want to say the potential issue with this is that the `positions` gradients will be tracked even if we only want to train on energies. Ideally we can fix in this PR, but if it gets too complicated we can open an issue to track.

At the very least, you should make sure there are no memory leaks, but the best case scenario is to not have the overhead if we can get around it.

File: nvalchemi/models/pipeline.py [hunk 1]
```diff
@@ -934,6 +943,7 @@ def _build_autograd_group_output(
                         displacement=strain.displacement,
                         orig_cell=strain.cell_for_stress,
                         retain_graph=retain_graph,
```
Comment (laserkelvin) [inline nvalchemi/models/pipeline.py:945]: Do the `retain_graph` flags need to factor in `training` as well?

**Review by laserkelvin — COMMENTED** [review]
Some comments for your consideration - functionally I think I would be able to approve, but we should get these things addressed if possible.

File: nvalchemi/models/pipeline.py (hunk 1 above)
Comment (ys-teh) [reply nvalchemi/models/pipeline.py:945]: I think it's already account for since "effective_retain = retain_graph or training" is set in autograd_forces and autograd_stresses.
https://github.com/NVIDIA/nvalchemi-toolkit/blob/df001caa30a8494d79b3c5348f5c7fd151b51fbb/nvalchemi/models/_utils.py#L76

File: test/models/test_pipeline.py
Comment (ys-teh) [reply test/models/test_pipeline.py:1]: Added test including stress.

**Review by laserkelvin — APPROVED** [review]

**Comment by laserkelvin** [comment]
/ok to test [e0015d3](https://github.com/NVIDIA/nvalchemi-toolkit/pull/99/commits/e0015d3735ae378a2b587ec4dca6a0ab3e55c875)
